h2/src/main/org/h2/mvstore/tx/Transaction.java

```diff
@@ -211,6 +224,12 @@ void log(int mapId, Object key, VersionedValue oldValue) {
     void logUndo() {
         long currentState = statusAndLogId.decrementAndGet();
         long logId = getLogId(currentState);
+        if (logId == LOG_ID_MASK) {
```

Comment:
This check may need adjustment too, but it depends on resolution of (2) from previous comment.